src/environmentd/tests/sql.rs

```diff
@@ -3909,3 +3909,74 @@ async fn test_builtin_schemas() {
         );
     }
 }
+
+// Test that DDLs are queued and processed serially.
+#[mz_ore::test(tokio::test(flavor = "multi_thread", worker_threads = 1))]
+#[cfg_attr(miri, ignore)] // too slow
+async fn test_serialized_ddl() {
+    let server = test_util::TestHarness::default().start().await;
+    server
+        .enable_feature_flags(&["enable_unsafe_functions"])
+        .await;
+    server
+        .enable_feature_flags(&["enable_unstable_dependencies"])
+        .await;
+    let client1 = server.connect().await.unwrap();
+    let client2 = server.connect().await.unwrap();
+    let client3 = server.connect().await.unwrap();
+    let cancel1 = client1.cancel_token();
+    let cancel3 = client3.cancel_token();
+
+    let handle = task::spawn(|| "test", async move {
+        // Wait hopefully enough time for the other create to start. This could flake if 100ms isn't
+        // enough time in CI, in which case increase this time.
+        tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
This seems like it will be very difficult to maintain. I don't think we need to explicitly test that DDL is executed serially, just that the behavior identically to some serial order. Here's an idea for a test:
1. Start with some database state `s`.
2. Generate `n` DDL statements that are valid on `s`, but each create a state `s'` where no other DDL is valid on `s'`.
3. Assert that `1` DDL statement was successful and `n-1` statements errored.

For example the following psuedo code would accomplish that.

```Rust
let mut handles = Vec::new();
let n = 10;

for i in 0..n {
    let handle = task::spawn(|| execute(format!("CREATE VIEW v AS SELECT {i}")));
    handles.push(handle);
}

let mut successes = 0;
let mut errors = 0;
for handle in handles {
    let result = handle.join();
    match result {
        Ok(_) => successes += 1;
        Err(_) => errors += 1;
    }
}

assert_eq!(successes, 1);
assert_eq!(errors, n-1);
```

That's the simplest example I could think of, but if we wanted to we could make this more complicated. The general idea being, come up with some DDL that we run concurrently where there is an enumerable number of valid end states. Then we can assert that the end state is one of the valid states.

Reply:
Good idea, done.